File: src/routes/habit.py
"""
LAWA 习惯引擎 — API 路由

提供：
  - POST /api/v1/habit/action      — 记录微行为
  - GET  /api/v1/habit/summary     — 今日行为摘要
  - GET  /api/v1/habit/habits      — 可用行为列表
  - GET  /api/v1/habit/feed        — 获取信息流
  - GET  /api/v1/habit/rewards     — 最近奖励
  - POST /api/v1/habit/config      — 更新配置
  - GET  /api/v1/habit/garden      — 花园状态
  - GET  /api/v1/habit/milestones  — 里程碑
"""
from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel, Field
from typing import Optional
from loguru import logger

from src.engine.action_engine import ActionEngine
from src.engine.reward_engine import RewardEngine
from src.engine.trigger_engine import TriggerEngine
from src.engine.investment_engine import InvestmentEngine

router = APIRouter(prefix="/api/v1/habit", tags=["habit"])


# ── 依赖注入 ──

def get_action_engine() -> ActionEngine:
    return ActionEngine()

def get_reward_engine() -> RewardEngine:
    return RewardEngine()

def get_trigger_engine() -> TriggerEngine:
    return TriggerEngine()

def get_investment_engine() -> InvestmentEngine:
    return InvestmentEngine()

async def get_current_user_id(user_id: str = "default_user") -> str:
    """从请求参数获取 user_id"""
    return user_id


# ── 请求/响应模型 ──

class RecordActionRequest(BaseModel):
    habit_code: str = Field(..., description="行为类型")
    duration_seconds: int = Field(0, description="耗时（秒）")
    completion_status: str = Field("completed", description="完成状态")
    triggered_by: str = Field("manual", description="触发方式")
    feed_id: Optional[str] = Field(None, description="关联信息流")

class UpdateComprehensionRequest(BaseModel):
    feed_id: str
    score: float = Field(..., ge=0, le=1)
    duration_seconds: int

class UpdateConfigRequest(BaseModel):
    trigger_time_slot: Optional[str] = None
    info_source_prefs: Optional[list[str]] = None
    action_prefs: Optional[list[str]] = None
    reward_frequency: Optional[str] = None
    feed_enabled: Optional[bool] = None
    morning_time: Optional[str] = None
    noon_time: Optional[str] = None
    evening_time: Optional[str] = None


class UpdateSocialLevelRequest(BaseModel):
    vocab_id: str
    new_level: str  # understand|use|create


# ── 路由 ──

@router.post("/action")
async def record_action(
    req: RecordActionRequest,
    engine: ActionEngine = Depends(get_action_engine),
    user_id: str = Depends(get_current_user_id),
):
    """记录一个微行为"""
    try:
        result = await engine.record_habit(
            user_id=user_id,
            habit_code=req.habit_code,
            duration_seconds=req.duration_seconds,
            completion_status=req.completion_status,
            triggered_by=req.triggered_by,
            feed_id=req.feed_id,
        )
        return {"status": "ok", "data": result}
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.get("/summary")
async def get_summary(
    engine: ActionEngine = Depends(get_action_engine),
    user_id: str = Depends(get_current_user_id),
):
    """今日行为摘要"""
    result = await engine.get_today_summary(user_id)
    return {"status": "ok", "data": result}


@router.get("/habits")
async def list_habits(
    engine: ActionEngine = Depends(get_action_engine),
):
    """列出所有可用行为类型"""
    return {"status": "ok", "data": engine.list_available_habits()}


@router.get("/feed")
async def get_feed(
    time_slot: str = "morning",
    engine: TriggerEngine = Depends(get_trigger_engine),
    user_id: str = Depends(get_current_user_id),
):
    """获取一条信息流内容"""
    result = await engine.get_feed(user_id, time_slot)
    return {"status": "ok", "data": result}


@router.get("/feeds")
async def list_feeds(
    limit: int = 10,
    engine: TriggerEngine = Depends(get_trigger_engine),
    user_id: str = Depends(get_current_user_id),
):
    """获取最近的信息流"""
    result = await engine.get_recent_feeds(user_id, limit)
    return {"status": "ok", "data": result}


@router.post("/feed/comprehension")
async def update_comprehension(
    req: UpdateComprehensionRequest,
    engine: TriggerEngine = Depends(get_trigger_engine),
    user_id: str = Depends(get_current_user_id),
):
    """更新信息流理解度评分"""
    ok = await engine.update_comprehension(user_id, req.feed_id, req.score, req.duration_seconds)
    if not ok:
        raise HTTPException(status_code=404, detail="Feed not found")
    return {"status": "ok"}


@router.get("/rewards")
async def list_rewards(
    limit: int = 10,
    engine: RewardEngine = Depends(get_reward_engine),
    user_id: str = Depends(get_current_user_id),
):
    """获取最近的可变奖励"""
    result = await engine.get_recent_rewards(user_id, limit)
    return {"status": "ok", "data": result}


@router.get("/config")
async def get_config(
    engine: TriggerEngine = Depends(get_trigger_engine),
    user_id: str = Depends(get_current_user_id),
):
    """获取用户习惯配置"""
    config = await engine.get_user_config(user_id)
    return {"status": "ok", "data": config}


@router.post("/config")
async def update_config(
    req: UpdateConfigRequest,
    engine: TriggerEngine = Depends(get_trigger_engine),
    user_id: str = Depends(get_current_user_id),
):
    """更新用户习惯配置"""
    updates = {k: v for k, v in req.model_dump().items() if v is not None}
    result = await engine.update_user_config(user_id, updates)
    return {"status": "ok", "data": result}


@router.get("/garden")
async def get_garden(
    engine: InvestmentEngine = Depends(get_investment_engine),
    user_id: str = Depends(get_current_user_id),
):
    """获取语言花园状态"""
    garden = await engine.get_garden_status(user_id)
    milestones = await engine.check_milestones(user_id)
    return {
        "status": "ok",
        "data": {
            "garden": garden,
            "new_milestones": milestones,
        },
    }


# ── 社交场景路由 ──

@router.get("/social/scene")
async def get_social_scene(
    category: str = "net_slang",
    lang_direction: str = "zh",
    engine: TriggerEngine = Depends(get_trigger_engine),
    user_id: str = Depends(get_current_user_id),
):
    """获取一条社交场景内容"""
    result = await engine.get_social_scene(user_id, lang_direction, category)
    return {"status": "ok", "data": result}


@router.get("/social/adaptive")
async def get_adaptive_scene(
    lang_direction: str = "zh",
    engine: TriggerEngine = Depends(get_trigger_engine),
    user_id: str = Depends(get_current_user_id),
):
    """根据用户社交理解度获取合适场景"""
    result = await engine.get_social_scene_by_level(user_id, lang_direction)
    return {"status": "ok", "data": result}


@router.post("/social/level")
async def update_social_level(
    req: UpdateSocialLevelRequest,
    engine: TriggerEngine = Depends(get_trigger_engine),
    user_id: str = Depends(get_current_user_id),
):
    """更新社交词汇理解等级"""
    ok = await engine.update_social_understanding(
        user_id, req.vocab_id, req.new_level
    )
    if not ok:
        raise HTTPException(status_code=404, detail="Vocab not found")
    return {"status": "ok"}


@router.get("/social/progress")
async def get_social_progress(
    engine: TriggerEngine = Depends(get_trigger_engine),
    user_id: str = Depends(get_current_user_id),
):
    """获取社交学习进度"""
    result = await engine.get_social_progress(user_id)
    return {"status": "ok", "data": result}


@router.get("/milestones")
async def list_milestones(
    engine: InvestmentEngine = Depends(get_investment_engine),
    user_id: str = Depends(get_current_user_id),
):
    """获取所有里程碑"""
    result = await engine.get_milestones(user_id)
    return {"status": "ok", "data": result}


@router.post("/asset")
async def add_asset(
    asset_type: str,
    asset_data: dict,
    engine: InvestmentEngine = Depends(get_investment_engine),
    user_id: str = Depends(get_current_user_id),
):
    """添加一条语言资产"""
    result = await engine.add_asset(user_id, asset_type, asset_data)
    return {"status": "ok", "data": result}


@router.get("/assets")
async def list_assets(
    asset_type: Optional[str] = None,
    limit: int = 20,
    engine: InvestmentEngine = Depends(get_investment_engine),
    user_id: str = Depends(get_current_user_id),
):
    """获取语言资产列表"""
    result = await engine.get_assets(user_id, asset_type, limit)
    return {"status": "ok", "data": result}


@router.get("/garden/vocab-details")
async def get_garden_vocab_details(
    engine: InvestmentEngine = Depends(get_investment_engine),
    user_id: str = Depends(get_current_user_id),
):
    """获取词汇明细（含社交词汇分类统计）"""
    result = await engine.get_vocab_details(user_id)
    return {"status": "ok", "data": result}


@router.get("/garden/growth")
async def get_garden_growth(
    engine: InvestmentEngine = Depends(get_investment_engine),
    user_id: str = Depends(get_current_user_id),
):
    """获取成长曲线数据"""
    result = await engine.get_growth_curve(user_id)
    return {"status": "ok", "data": result}


@router.get("/garden/report")
async def get_garden_report(
    engine: InvestmentEngine = Depends(get_investment_engine),
    user_id: str = Depends(get_current_user_id),
):
    """获取花园周报"""
    result = await engine.get_garden_report(user_id)
    return {"status": "ok", "data": result}


@router.get("/garden/health")
async def get_health_insights(
    engine: InvestmentEngine = Depends(get_investment_engine),
    user_id: str = Depends(get_current_user_id),
):
    """获取习惯健康度洞察"""
    result = await engine.get_health_insights(user_id)
    return {"status": "ok", "data": result}

File: frontend/src/api/habit.ts
/**
 * LAWA2 — 习惯引擎 API 封装
 *
 * 所有后端响应格式: { status: "ok", data: ... }
 * 此层负责解包，各页面直接消费 data
 * user_id 从 session store 自动获取
 */
const BASE = '/api/v1/habit'

interface ApiResponse<T> {
  status: string
  data: T
}

function getUserId(): string {
  // 惰性加载避免循环依赖
  try {
    const raw = localStorage.getItem('lawa2_session')
    if (!raw) return 'default_user'
    const s = JSON.parse(raw)
    return s?.userId || 'default_user'
  } catch {
    return 'default_user'
  }
}

async function fetchData<T>(url: string, init?: RequestInit): Promise<T> {
  const res = await fetch(url, {
    headers: { 'Content-Type': 'application/json' },
    ...init,
  })
  if (!res.ok) {
    const err = await res.json().catch(() => ({ detail: res.statusText }))
    throw new Error(err.detail || `HTTP ${res.status}`)
  }
  const json: ApiResponse<T> = await res.json()
  return json.data
}

function withUser(path: string): string {
  const uid = getUserId()
  const sep = path.includes('?') ? '&' : '?'
  return `${BASE}${path}${sep}user_id=${encodeURIComponent(uid)}`
}

// ── 数据模型 ──

export interface HabitConfig {
  trigger_time_slot: string
  info_source_prefs: string[]
  action_prefs: string[]
  reward_frequency: string
  feed_enabled: boolean
  morning_time: string
  noon_time: string
  evening_time: string
  total_xp: number
  current_streak: number
  longest_streak: number
  total_actions: number
  today_actions: number
}

export interface FeedData {
  feed_id: string
  content_type: string
  text: string
  difficulty: string
  vocab_hints: string[]
  source: string
  is_first_today: boolean
  reward: FeedReward | null
}

export interface FeedReward {
  id: string
  reward_type: string
  xp_bonus: number
  surprise_level: number
  reward_value: {
    type: string
    name: string
    icon: string
    message: string
  }
}

export interface ActionResponse {
  habit_log_id: string
  xp_earned: number
  streak_days: number
  total_habits_today: number
  reward: ActionReward | null
}

export interface ActionReward {
  id: string
  reward_type: string
  xp_bonus: number
  reward_value: {
    type: string
    name: string
    icon: string
    message: string
  }
}

export interface SummaryData {
  total_habits: number
  total_xp: number
  by_type: Record<string, number>
  streak_days: number
}

export interface GardenData {
  garden: {
    total_vocab: number
    total_assets: number
    total_milestones: number
    streak_days: number
    habit_level: number
    total_xp: number
    by_stage: Record<string, number>
  }
  new_milestones: Array<{
    code: string
    name: string
    description: string
    celebration_type: string
  }>
}

export interface RewardData {
  id: string
  reward_type: string
  xp_bonus: number
  surprise_level: number
  reward_value: {
    type: string
    name: string
    icon: string
    message: string
  }
  created_at: string
}

export interface SocialSceneData {
  vocab_id: string
  word: string
  meaning: string
  scene_example: string
  difficulty: string
  category: string
  lang_direction: string
  understanding_level: string
}

export interface SocialProgressData {
  by_category: Record<string, { understand: number; use: number; create: number }>
  total_understand: number
  total_use: number
  total_all: number
}

export interface MilestoneData {
  code: string
  name: string
  description: string
  category: string
  target_value: number
  is_achieved: boolean
  progress: number
  achieved_at: string | null
  celebration_type: string | null
  icon: string
}

export interface RecordActionBody {
  habit_code: string
  duration_seconds: number
  completion_status: string
  triggered_by: string
  feed_id?: string
}

// ── API 方法 ──

export async function getConfig(): Promise<HabitConfig> {
  return fetchData(withUser('/config'))
}

export async function getFeed(): Promise<FeedData> {
  return fetchData(withUser('/feed'))
}

export async function recordAction(body: RecordActionBody): Promise<ActionResponse> {
  return fetchData(withUser('/action'), {
    method: 'POST',
    body: JSON.stringify(body),
  })
}

export async function getSummary(): Promise<SummaryData> {
  return fetchData(withUser('/summary'))
}

export async function getGarden(): Promise<GardenData> {
  return fetchData(withUser('/garden'))
}

export async function getRewards(): Promise<RewardData[]> {
  return fetchData(withUser('/rewards'))
}

export async function getMilestones(): Promise<MilestoneData[]> {
  return fetchData(withUser('/milestones'))
}

export interface SocialSceneData {
  vocab_id: string
  word: string
  meaning: string
  scene_example: string
  difficulty: string
  category: string
  lang_direction: string
  understanding_level: string
}

export interface SocialProgressData {
  by_category: Record<string, { understand: number; use: number; create: number }>
  total_understand: number
  total_use: number
  total_all: number
}

export async function getSocialScene(langDirection: string = 'zh'): Promise<SocialSceneData> {
  return fetchData(`${withUser('/social/adaptive')}&lang_direction=${langDirection}`)
}

export async function updateSocialLevel(vocabId: string, newLevel: string): Promise<void> {
  await fetchData(`${withUser('/social/level')}`, {
    method: 'POST',
    body: JSON.stringify({ vocab_id: vocabId, new_level: newLevel }),
  })
}

export async function getSocialProgress(): Promise<SocialProgressData> {
  return fetchData(withUser('/social/progress'))
}

export async function updateComprehension(
  feed_id: string,
  score: number,
  duration_seconds: number
): Promise<void> {
  await fetchData(withUser('/feed/comprehension'), {
    method: 'POST',
    body: JSON.stringify({ feed_id, score, duration_seconds }),
  })
}

export interface GardenReportData {
  report: string
  highlights: string[]
  week_actions: number
  week_vocab: number
  total_xp: number
  level: number
  streak: number
  can_water: boolean
  watered_today: boolean
  lang_direction: string
}

export interface HealthInsightsData {
  health_score: number
  trend: 'up' | 'down' | 'stable'
  dimensions: {
    consistency: number
    depth: number
    breadth: number
    recovery: number
  }
  recommendation: string
}

export async function getGardenReport(): Promise<GardenReportData> {
  return fetchData(withUser('/garden/report'))
}

export async function getHealthInsights(): Promise<HealthInsightsData> {
  return fetchData(withUser('/garden/health'))
}